File: ios/oplayer/bitshares/GrapheneWebSocket.m
//
//  GrapheneWebSocket.m
//
//  Created by SYALON on 13-9-3.
//
//

#import "GrapheneWebSocket.h"
#import "OrgUtils.h"

#import <Crashlytics/Crashlytics.h>
#import <Flurry/Flurry.h>

#define kGwsKeepAliveINterval   5000
#define kGwsMaxSendLife         4
#define kGwsMaxRecvLife         10
#define kGwsConnectPromiseKey   @"__connectPromiseCallback"

@interface GrapheneWebSocket()
{
    SRWebSocket*                _webSocket;
    NSString*                   _url;
    KeepAliveCallback           _keepAliveCb;
    
    __weak WsPromise*           __weak_connect_promise;
    BOOL                        _closed;
    
    NSUInteger                  _cbId;
    NSMutableDictionary*        _cbs;       //  普通请求的回调函数列表
    NSMutableDictionary*        _subs;      //  订阅推送回调列表
    NSMutableDictionary*        _unsub;     //  取消订阅
    
    NSTimer*                    _keepAliveTimer;
    NSInteger                   _send_life;
    NSInteger                   _recv_life;
}
@end

@implementation GrapheneWebSocket

/**
 *  (private) 处理连接超时
 */
- (void)onConnectTimeoutTrigger:(NSInteger)connectTimeoutSec
{
    //  连接成功，则连接的promise回调已经被移除了。
    id promise_callbacks = [_cbs objectForKey:kGwsConnectPromiseKey];
    if (!promise_callbacks){
        return;
    }
    
    //  连接失败
    WsRejectHandler reject = [promise_callbacks objectForKey:@"reject"];
    [_cbs removeObjectForKey:kGwsConnectPromiseKey];
    [self close_websocket];
    reject([WsPromiseException makeException:[NSString stringWithFormat:@"Connection attempt timed out after %@s", @(connectTimeoutSec)]]);
}

- (void)dealloc
{
    _url = nil;
    _keepAliveCb = nil;
    __weak_connect_promise = nil;
    
    _cbs = nil;
    _subs = nil;
    [self close_websocket];
}

- (id)initWithServer:(NSString*)url connect_timeout:(NSInteger)connect_timeout keepaliveCb:(KeepAliveCallback)keepaliveCb
{
    self = [super init];
    if (self)
    {
        _webSocket = nil;
        _url = url;
        _keepAliveCb = keepaliveCb;
        _closed = NO;
        
        _cbId = 0;
        _cbs = [NSMutableDictionary dictionary];
        _subs = [NSMutableDictionary dictionary];
        _unsub = [NSMutableDictionary dictionary];
        
        _keepAliveTimer = nil;
        _send_life = kGwsMaxSendLife;
        _recv_life = kGwsMaxRecvLife;
        
        //  --- 开始初始化 ---
        __weak_connect_promise = [WsPromise promise:^(WsResolveHandler resolve, WsRejectHandler reject) {
            //  保存 resolve 和 reject（这里cbs等会对promise保留引用，因此self里用weak即可，不然会循环引用。
            [_cbs setObject:@{@"resolve":resolve, @"reject":reject} forKey:kGwsConnectPromiseKey];
            //  连接socket
            _webSocket = [[SRWebSocket alloc] initWithURLRequest:[NSURLRequest requestWithURL:[NSURL URLWithString:url]]];
            _webSocket.delegate = self;
            [_webSocket open];
        }];
        
        //  连接超时定时器
        __weak id this = self;
        dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(connect_timeout * NSEC_PER_SEC)), dispatch_get_main_queue(), ^{
            if (this){
                [this onConnectTimeoutTrigger:connect_timeout];
            }
        });
    }
    return self;

}

- (WsPromise*)login:(NSString*)username password:(NSString*)password
{
    assert(__weak_connect_promise);
    return [__weak_connect_promise then:(^id(id data) {
        return [self call:@[@(1), @"login", @[username, password]]];
    })];
}

- (WsPromise*)call:(NSArray*)params
{
    //  已经断线
    if (_closed){
        return [WsPromise reject:[WsPromiseException makeException:@"连接已关闭..."]];
    }
    
    //  ID计数器
    _cbId++;
    
    //  部分方法特殊处理
    NSString* method = [params objectAtIndex:1];
    if ([method isEqualToString:@"set_subscribe_callback"] ||
        [method isEqualToString:@"subscribe_to_market"] ||
        [method isEqualToString:@"broadcast_transaction_with_callback"] ||
        [method isEqualToString:@"set_pending_transaction_callback"])
    {
        id old_sub_params = [params objectAtIndex:2];
        id sub_callback = [old_sub_params firstObject];
        
        //  订阅的callback替换为cbId传送到服务器。
        id sub_params_mutable = [old_sub_params mutableCopy];
        [sub_params_mutable replaceObjectAtIndex:0 withObject:@(_cbId)];
        id new_sub_params = [sub_params_mutable copy];
        
        //  保存订阅callback
        [_subs setObject:sub_callback forKey:@(_cbId)];
        
        //  更新订阅接口的参数信息
        id new_params = [params mutableCopy];
        [new_params replaceObjectAtIndex:2 withObject:new_sub_params];
        params = [new_params copy];
    }
    
    //  取消订阅
    if ([method isEqualToString:@"unsubscribe_from_market"] ||
        [method isEqualToString:@"unsubscribe_from_accounts"])
    {
        id old_sub_params = [params objectAtIndex:2];
        id unsub_callback = [old_sub_params firstObject];
        for (id key in _subs) {
            id cb = [_subs objectForKey:key];
            if ([cb isEqual:unsub_callback]){
                [_unsub setObject:key forKey:@(_cbId)];
                break;
            }
        }
        
        //  移除第一个 callback 参数
        id sub_params_mutable = [old_sub_params mutableCopy];
        [sub_params_mutable removeObjectAtIndex:0];
        id new_params = [params mutableCopy];
        [new_params replaceObjectAtIndex:2 withObject:[sub_params_mutable copy]];
        params = [new_params copy];
    }
    
    //  TODO:fowallet 取消订阅 unsubscribe_from_market unsubscribe_from_accounts
    NSLog(@"[ApiCall] %@", method);
    
    //  序列化
    NSError* err = nil;
    NSData* data = [NSJSONSerialization dataWithJSONObject:@{@"id":@(_cbId), @"method":@"call", @"params":params}
                                                   options:NSJSONReadingAllowFragments
                                                     error:&err];
    if (err || !data){
        return [WsPromise reject:[WsPromiseException makeException:err]];
    }
    _send_life = kGwsMaxSendLife;

    //  构造promise对象并发送数据
    WsPromise* p = [WsPromise promise:^(WsResolveHandler resolve, WsRejectHandler reject) {
        NSError* error = nil;
        BOOL ret = [_webSocket sendData:data error:&error];
        if (!ret || error){
            reject([WsPromiseException makeException:error]);
        }else{
            [_cbs setObject:@{@"resolve":resolve, @"reject":reject} forKey:@(_cbId)];
        }
    }];
    return p;
}

/**
 *  (private) 处理服务器响应
 */
- (void)listener:(id)response
{
    assert(response);
    
    id callback_id = [response objectForKey:@"id"];
    BOOL sub = NO;
    
    //  {"method":"notice","params":[notify_id,notify_data...]]}
    id method = [response objectForKey:@"method"];
    if (method && [method isEqualToString:@"notice"]){
        sub = YES;
        callback_id = [[response objectForKey:@"params"] firstObject];
    }
    
    id callback;
    if (sub){
        callback = [_subs objectForKey:callback_id];
    }else{
        callback = [_cbs objectForKey:callback_id];
    }
    
    if (callback && sub)
    {
        //  订阅方法 callback
        WsNotifyCallback cb = (WsNotifyCallback)callback;
        if (cb(YES, [[response objectForKey:@"params"] objectAtIndex:1])){
            //  移除 callback
            [_subs removeObjectForKey:callback_id];
        }
    }
    else if (callback && !sub)
    {
        //  普通请求 callback
        [_cbs removeObjectForKey:callback_id];   //  remove
        
        //  API调用是否异常判断
        id resp_error = [response objectForKey:@"error"];
        if (resp_error){
            //  TODO:fowallet 错误分类统计
            
            //  错误格式：
            //    @{"code":"1", @"data":@{@"code":@"code", @"message":@"base error message", @"name":@"xx", @"stack":@{}}, @"message":@"detail message"}
            //  [统计]
            if ([resp_error isKindOfClass:[NSDictionary class]]){
                id detail_error_message = [resp_error objectForKey:@"message"];
                id error_data = [resp_error objectForKey:@"data"];
                if (error_data){
                    id error_message = [error_data objectForKey:@"message"];
                    id error_stack = [error_data objectForKey:@"stack"];
                    id error_code = [error_data objectForKey:@"code"];
                    if (error_message && error_code && error_stack && [error_stack count] > 0){
                        
                        NSError* log_err = nil;
                        NSString* log_str = nil;
                        NSData* log_data = [NSJSONSerialization dataWithJSONObject:resp_error
                                                                           options:NSJSONReadingAllowFragments
                                                                             error:&log_err];
                        if (!log_err && log_data){
                            log_str = [[NSString alloc] initWithData:log_data encoding:NSUTF8StringEncoding];
                        }
                        [OrgUtils logEvents:[NSString stringWithFormat:@"api_error_%@", error_code]
                                     params:@{@"last_stack":[error_stack lastObject],
                                              @"message":error_message,
                                              @"detail_message":detail_error_message ? : @"",
                                              @"error":log_str ? : @""
                                              }];
                    }
                }
                if (detail_error_message){
                    CLS_LOG(@"graphene api error: %@", detail_error_message);
                }
            }
            //  resp_error 可能为 {message, data, code} 格式的 Hash。
            WsResolveHandler reject = [callback objectForKey:@"reject"];
            reject([WsPromiseException makeException:resp_error]);
        }else{
            WsResolveHandler resolve = [callback objectForKey:@"resolve"];
            resolve([response objectForKey:@"result"]);
        }
        //  取消订阅
        id unsub_id = [_unsub objectForKey:callback_id];
        if (unsub_id){
            [_subs removeObjectForKey:unsub_id];
            [_unsub removeObjectForKey:callback_id];
        }
    }
    else
    {
        NSLog(@"Warning: unknown websocket response: %@", response);
    }
}

- (void)close_websocket
{
    if (_webSocket){
        _webSocket.delegate = nil;
        [_webSocket close];
        _webSocket = nil;
    }
}

- (void)close
{
    [self processWebsocketErrorOrClose:@"user close..."];
}

- (BOOL)isClosed
{
    return _closed;
}

#pragma mark- for keepAliveTimer
- (void)startKeepAliveTimer
{
    if (!_keepAliveTimer){
        _keepAliveTimer = [NSTimer scheduledTimerWithTimeInterval:5.0
                                                           target:self
                                                         selector:@selector(onKeepAliveTimerTick)
                                                         userInfo:nil
                                                          repeats:YES];
        [_keepAliveTimer fire];
    }
}

- (void)stopKeepAliveTimer
{
    if (_keepAliveTimer){
        [_keepAliveTimer invalidate];
        _keepAliveTimer = nil;
    }
}

- (void)onKeepAliveTimerTick
{
    --_recv_life;
    if (_recv_life <= 0){
        NSLog(@"%@ connection is dead, terminating ws", _url);
        [self processWebsocketErrorOrClose:@"heartbeat..."];
        return;
    }
    
    --_send_life;
    if (_send_life <= 0){
        if (_keepAliveCb){
            _keepAliveCb(_closed);
        }
        _send_life = kGwsMaxSendLife;
    }
}

#pragma mark- for SRWebSocketDelegate

/**
 Called when any message was received from a web socket.
 This method is suboptimal and might be deprecated in a future release.
 
 @param webSocket An instance of `SRWebSocket` that received a message.
 @param message   Received message. Either a `String` or `NSData`.
 */
- (void)webSocket:(SRWebSocket *)webSocket didReceiveMessage:(id)message
{
    //  重置接受数据包的心跳计数
    _recv_life = kGwsMaxRecvLife;
    
    //  解析服务器响应为 json 格式
    NSData* data = nil;
    if ([message isKindOfClass:[NSString class]]){
        data = [message dataUsingEncoding:NSUTF8StringEncoding];
    }else{
        data = (NSData*)message;
    }
    NSError* err = nil;
    id response = [NSJSONSerialization JSONObjectWithData:data options:NSJSONReadingAllowFragments error:&err];
    if (err || !response){
        [self processWebsocketErrorOrClose:err];
        return;
    }
    
    //  处理响应
    [self listener:response];
}

/**
 Called when a given web socket was open and authenticated.
 
 @param webSocket An instance of `SRWebSocket` that was open.
 */
- (void)webSocketDidOpen:(SRWebSocket *)webSocket
{
    //  REMARK：如果刚连接成功就已经超时了，则该callback可能为空。
    id promise_callbacks = [_cbs objectForKey:kGwsConnectPromiseKey];
    if (!promise_callbacks){
        return;
    }
    
    WsResolveHandler resolve = [promise_callbacks objectForKey:@"resolve"];
    [_cbs removeObjectForKey:kGwsConnectPromiseKey];
    
    //  REMARK：心跳定时器
    [self startKeepAliveTimer];
    
    //  连接成功
    resolve(@"websocket opened...");
}

/**
 *  (private) 处理网络异常 or 网络关闭。
 */
- (void)processWebsocketErrorOrClose:(id)error
{
    if (!_closed){
        //  关闭标记
        _closed = YES;
        
        //  promise_callbacks 存在则为连接异常，否则是通讯异常。
        id promise_callbacks = [_cbs objectForKey:kGwsConnectPromiseKey];
        if (promise_callbacks){
            WsRejectHandler reject = [promise_callbacks objectForKey:@"reject"];
            [_cbs removeObjectForKey:kGwsConnectPromiseKey];
            reject([WsPromiseException makeException:error]);
        }
        
        //  取消心跳计时器
        [self stopKeepAliveTimer];
        
        //  通讯中异常：则当前的所有待完成promise全部reject
        if ([_cbs count] > 0){
            for(id key in _cbs) {
                id cb = [_cbs objectForKey:key];
                WsResolveHandler reject = [cb objectForKey:@"reject"];
                reject([WsPromiseException makeException:@"websocket closed..."]);
            }
            [_cbs removeAllObjects];
        }
        //  当前订阅的callback也全部push false。
        if ([_subs count] > 0){
            for (id key in _subs) {
                WsNotifyCallback cb = (WsNotifyCallback)[_subs objectForKey:key];
                //  NOTIFY: NO，网络中断。
                cb(NO, nil);
            }
            [_subs removeAllObjects];
        }
        [_unsub removeAllObjects];
        
        //  关闭网络连接
        [self close_websocket];
    }
}

/**
 Called when a given web socket encountered an error.
 
 @param webSocket An instance of `SRWebSocket` that failed with an error.
 @param error     An instance of `NSError`.
 */
- (void)webSocket:(SRWebSocket *)webSocket didFailWithError:(NSError *)error
{
    NSLog(@"websocket: didFailWithError: %@", error);
    [self processWebsocketErrorOrClose:error];
}

/**
 Called when a given web socket was closed.
 
 @param webSocket An instance of `SRWebSocket` that was closed.
 @param code      Code reported by the server.
 @param reason    Reason in a form of a String that was reported by the server or `nil`.
 @param wasClean  Boolean value indicating whether a socket was closed in a clean state.
 */
- (void)webSocket:(SRWebSocket *)webSocket didCloseWithCode:(NSInteger)code reason:(nullable NSString *)reason wasClean:(BOOL)wasClean
{
    NSLog(@"websocket: didCloseWithCode: code: %@ reason: %@", @(code), reason);
    [self processWebsocketErrorOrClose:@"websocket closed..."];
}

@end

File: android/app/src/main/java/bitshares/GrapheneWebSocket.kt
package com.fowallet.walletcore.bts

import bitshares.*
import com.crashlytics.android.Crashlytics
import org.json.JSONArray
import org.json.JSONObject
import java.net.URI
import java.util.*

/**
 * websocket status
 */
enum class GrapheneSocketStatus(val value: Int) {
    pending(0),                 //  连接中
    logined(1),                 //  已连接
    closed(2),                  //  断开连接
}

const val kGwsKeepAliveINterval: Int = 5000
const val kGwsMaxSendLife: Int = 5
const val kGwsMaxRecvLife: Int = kGwsMaxSendLife * 2
const val kGwsConnectPromiseKey: String = "__connectPromiseCallback"

class GrapheneWebSocket {
    private var _webSocket: BtsWsClient? = null

    lateinit var _keepaliveCb: (sock: GrapheneWebSocket) -> Unit
    private var _cbId: Int = 0
    private var _cbs: MutableMap<String, Promise> = mutableMapOf()      //  普通请求的回调函数列表
    private var _subs: MutableMap<String, Any> = mutableMapOf()         //  订阅推送回调列表
    private var _unsub: MutableMap<String, Any> = mutableMapOf()        //  取消订阅
    private var _keepAliveTimer: Timer? = null
    private var _send_life: Int = 0
    private var _recv_life: Int = 0

    //  --------------------------
    private var _username: String = ""
    private var _password: String = ""
    private var _auto_reconnect: Boolean = true
    private var _reconnect_times: Int = 0
    private var _url_list = mutableListOf<String>()
    private var _api_list: Array<String>? = null
    private var _api_ids = mutableMapOf<String, Number>()

    private var _curr_wsnode: String = ""
    private var _wait_promises_queue = mutableListOf<Promise>()
    private var _conn_promise: Promise? = null
    private var _status = GrapheneSocketStatus.pending

    /**
     * (public) 调用请求
     */
    fun call(api_name: String, method: String, params: JSONArray): Promise {
        return wait().then {
            val api_id = _api_ids[api_name]
            if (api_id != null) {
                return@then exec(jsonArrayfrom(api_id, method, params))
            } else {
                return@then Promise._reject("unsupported apiname: ${api_name}")
            }
        }
    }

    /**
     * 是否连接并且登录成功
     */
    fun is_connected(): Boolean {
        return _status == GrapheneSocketStatus.logined
    }

    /**
     * 是否断开连接
     */
    private fun is_closed(): Boolean {
        return _status == GrapheneSocketStatus.closed
    }

    /**
     * 是否在连接中
     */
    fun is_pending(): Boolean {
        return _status == GrapheneSocketStatus.pending
    }

    /**
     * append a new node
     */
    fun add_node(url: String) {
        _url_list.add(url)
    }

    fun initWithServer(url: String, api_list: Array<String>, connect_timeout: Int, keepaliveCb: (sock: GrapheneWebSocket) -> Unit): GrapheneWebSocket {
        _url_list.add(url)

        _webSocket = null
        _keepaliveCb = keepaliveCb
        _api_list = api_list
        _auto_reconnect = true

        //  --- 开始初始化 ---
        reconnect()
        return this
    }

    /**
     * reconnect
     */
    fun manual_reconnect() {
        if (is_closed()) {
            reconnect()
        }
    }

    /**
     * (private) 获取连接的节点地址（根据数组大小进行轮询）
     */
    private fun gen_next_ws_node(): String {
        assert(_url_list.isNotEmpty())
        return _url_list[_reconnect_times++ % _url_list.size]
    }

    /**
     * (private) 开始连接
     */
    private fun reconnect() {
        //  gen url
        _curr_wsnode = gen_next_ws_node()

        //  当前请求数据
        _cbId = 0
        _cbs = mutableMapOf()
        _subs = mutableMapOf()
        _unsub = mutableMapOf()

        //  连接等待队列
        _wait_promises_queue = mutableListOf<Promise>()
        _api_ids = mutableMapOf<String, Number>()

        //  心跳数据
        _keepAliveTimer = null
        _send_life = kGwsMaxSendLife
        _recv_life = kGwsMaxRecvLife

        //  status
        _status = GrapheneSocketStatus.pending

        //  生成等待连接promise
        _conn_promise = Promise()
        _conn_promise!!.then {
            on_connect_responsed(it, null)
            return@then null
        }.catch {
            on_connect_responsed(null, it)
        }

        //  初始化websocket连接
        _webSocket = BtsWsClient(URI(_curr_wsnode), this, 5000)
        _webSocket!!.connect()
    }

    /**
     * (private) 事件 - 连接成功 or 连接失败
     */
    private fun on_connect_responsed(data: Any?, error: Any?) {
        if (error != null) {
            _status = GrapheneSocketStatus.closed
            for (promise in _wait_promises_queue) {
                promise.reject(error)
            }
            _wait_promises_queue.clear()
        } else {
            if (is_closed()) {
                on_login_responsed(null, "closed on pending...")
            } else {
                exec(jsonArrayfrom(1, "login", jsonArrayfrom(_username, _password))).then {
                    on_login_responsed(it, null)
                    return@then null
                }.catch { err ->
                    on_login_responsed(null, err)
                }
            }
        }
    }

    /**
     * (private) 事件 - 登录成功 or 登录失败
     */
    private fun on_login_responsed(data: Any?, error: Any?) {
        if (error != null) {
            _status = GrapheneSocketStatus.closed
            for (promise in _wait_promises_queue) {
                promise.reject(error)
            }
            _wait_promises_queue.clear()
        } else {
            val promise_list = JSONArray()
            _api_list!!.forEach {
                promise_list.put(exec(jsonArrayfrom(1, it, JSONArray())))
            }
            Promise.all(promise_list).then {
                on_api_init_responsed(it as? JSONArray, null)
                return@then null
            }.catch { err ->
                on_api_init_responsed(null, err)
            }
        }
    }

    /**
     * (private) 事件 - 初始化APIID 成功 or 失败
     */
    private fun on_api_init_responsed(data_array: JSONArray?, error: Any?) {
        if (error != null) {
            _status = GrapheneSocketStatus.closed
            for (promise in _wait_promises_queue) {
                promise.reject(error)
            }
            _wait_promises_queue.clear()
        } else {
            var idx = 0
            for (api_id in data_array!!.forin<Number>()) {
                _api_ids[_api_list!![idx++]] = api_id!!
            }
            _status = GrapheneSocketStatus.logined
            for (promise in _wait_promises_queue) {
                promise.resolve(this)
            }
            _wait_promises_queue.clear()
        }
    }

    /**
     * (private) 等待连接成功
     */
    private fun wait(): Promise {
        if (is_connected()) {
            return Promise._resolve(this)
        }
        if (is_closed() && !_auto_reconnect) {
            return Promise._reject("websocket disconnected...")
        }
        if (is_closed()) {
            reconnect()
        }
        var p = Promise()
        _wait_promises_queue.add(p)
        return p
    }

    /**
     * (private) 处理请求
     */
    private fun exec(params: JSONArray): Promise {
        //  已经断线
        if (is_closed()) {
            return Promise._reject("websocket disconnected...")
        }

        //  TODO:该方法待check

        // 计算器
        _cbId++

        //  部分方法特殊处理
        val method: String = params[1] as String
        if (method == "set_subscribe_callback" ||
                method == "subscribe_to_market" ||
                method == "broadcast_transaction_with_callback" ||
                method == "set_pending_transaction_callback") {

            val old_sub_params = params[2] as JSONArray
            val sub_callback = old_sub_params[0]

            //  订阅的callback替换为cbId传送到服务器。
            old_sub_params.put(0, _cbId.toString())

            //  保存订阅callback
            _subs[_cbId.toString()] = sub_callback
        }

        //  取消订阅
        if (method == "unsubscribe_from_market" || method == "unsubscribe_from_accounts") {
            val old_sub_params = params[2] as JSONArray
            val unsub_callback = old_sub_params[0]
            _subs.keys.forEach {
                val cb = _subs[it]
                if (cb === unsub_callback) {
                    _unsub[_cbId.toString()] = it
                    return@forEach
                }
            }
            //  移除第一个 callback 参数
            val sub_params_mutable = JSONArray()
            var idx = 0
            for (obj in old_sub_params.forin<Any>()) {
                if (idx != 0) {
                    sub_params_mutable.put(obj)
                }
                ++idx
            }
            params.put(2, sub_params_mutable)
        }


        //  TODO:fowallet 取消订阅 unsubscribe_from_market unsubscribe_from_accounts

        //  序列化
        val data = jsonObjectfromKVS("id", _cbId, "method", "call", "params", params)

        //  构造promise对象并发送数据
        _send_life = kGwsMaxSendLife

        val p = Promise()
        try {
            val send_data = data.toString()
            _webSocket!!.send(send_data)
            assert(_cbs[_cbId.toString()] == null)
            _cbs[_cbId.toString()] = p
        } catch (e: Exception) {
            p.reject(e)
        }
        return p
    }

    private fun listener(response: JSONObject) {
        var callback_id = response.optString("id", "")

        var sub = false
        val method: String = response.optString("method")

        if (method == "notice") {
            sub = true
            callback_id = response.getJSONArray("params").getString(0)
        }

        var callback: Any? = null

        when (sub) {
            true -> callback = _subs[callback_id]
            false -> callback = _cbs[callback_id]
        }

        if (callback != null && sub) {

            //  订阅方法 callback
            val cb = callback as (Boolean, Any) -> Unit
            cb(true, response.getJSONArray("params")[1])

        } else if (callback != null && !sub) {

            //  普通请求 callback
            _cbs.remove(callback_id)

            //  API调用是否异常判断
            val resp_error = response.optJSONObject("error")
            callback = callback as Promise
            if (resp_error != null) {
                //  错误格式：
                //    @{"code":"1", @"data":@{@"code":@"code", @"message":@"base error message", @"name":@"xx", @"stack":@{}}, @"message":@"detail message"}
                //  [统计]
                val detail_error_message = resp_error.optString("message")
                //  统计Crash日志
                Crashlytics.log(detail_error_message)
                val error_data = resp_error.optJSONObject("data")
                if (error_data != null) {
                    val error_message = error_data.optString("message")
                    val error_code = error_data.optInt("code").toString()
                    val error_stack = error_data.optJSONArray("stack")
                    if (error_message != "" && error_code != "" && error_stack != null && error_stack.length() > 0) {
                        val stack_str = error_stack.last<JSONObject>().toString()
                        btsppLogCustom("api_error_${error_code}",
                                jsonObjectfromKVS("message", error_message, "detail_message", detail_error_message, "stack_str", stack_str))
                    }
                }
                //  返回
                callback.reject(resp_error)
            } else {
                callback.resolve(response.get("result"))
            }
            //  取消订阅
            val unsub_id = _unsub[callback_id]
            if (unsub_id != null) {
                _subs.remove(unsub_id)
                _unsub.remove(callback_id)
            }
        } else {
            assert(false)
        }
    }

    fun close() {
        process_websocket_error_or_close("user close...")
    }

    private fun startKeepAliveTimer() {
        if (_keepAliveTimer == null) {
            _keepAliveTimer = Timer()
            _keepAliveTimer!!.schedule(object : TimerTask() {
                override fun run() {
                    delay_main {
                        onKeepAliveTimerTick()
                    }
                }
            }, 5000, 5000)
        }
    }

    private fun stopKeepAliveTimer() {
        if (_keepAliveTimer != null) {
            _keepAliveTimer!!.cancel()
            _keepAliveTimer = null
        }
    }

    fun onKeepAliveTimerTick() {
        --_recv_life
        if (_recv_life <= 0) {
            process_websocket_error_or_close("heartbeat...")
            return
        }

        --_send_life
        if (_send_life <= 0) {
            if (_keepaliveCb != null) {
                _keepaliveCb(this)
            }
            _send_life = kGwsMaxSendLife
        }
    }


    /**
    Called when any message was received from a web socket.
    This method is suboptimal and might be deprecated in a future release.

    @param webSocket An instance of `SRWebSocket` that received a message.
    @param message   Received message. Either a `String` or `NSData`.
     */
    fun didReceiveMessage(message: Any) {
        //  重置接受数据包的心跳计数
        _recv_life = kGwsMaxRecvLife

        //  解析服务器响应为 json 格式
        var data: JSONObject? = null

        try {
            if (message::class == String::class) {
                data = JSONObject(message.toString().trim())
            } else {
                data = message as JSONObject
            }
        } catch (err: Exception) {
            process_websocket_error_or_close("websocket events message: invalid json, ${err.message}")
            return
        }
        listener(data)
    }

    /**
    Called when a given web socket was open and authenticated.

    @param webSocket An instance of `SRWebSocket` that was open.
     */
    fun webSocketDidOpen() {
        //  REMARK：心跳定时器
        startKeepAliveTimer()
        //  连接成功
        assert(_conn_promise != null)
        _conn_promise?.resolve("websocket opened...")
        _conn_promise = null
    }

    fun process_websocket_error_or_close(message: String) {
        if (is_closed()) {
            return
        }

        //  处于 pending、logined 状态则直接关闭
        _status = GrapheneSocketStatus.closed

        _conn_promise?.reject(message)
        _conn_promise = null

        //  取消心跳计时器
        stopKeepAliveTimer()

        //  通讯中或登录中异常：则当前的所有待完成promise全部reject
        if (_cbs.isNotEmpty()) {
            for ((_, v) in _cbs) {
                v.reject(message)
            }
            _cbs.clear()
        }
        if (_subs.isNotEmpty()) {
            for ((_, v) in _subs) {
                val cb = v as (Boolean, Any?) -> Unit
                cb.invoke(false, message)
            }
            _subs.clear()
        }

        //  关闭网络连接
        _webSocket?.close()
        _webSocket = null

        //  TODO:unse?
//        _unsub.clear()
    }

    fun didFailWithError(error: Exception) {
        process_websocket_error_or_close("websocket events error, ${error.message}")
    }

}


